capsules/src/segger_rtt.rs

```diff
@@ -0,0 +1,194 @@
+//! if there is support for software interrupts or deferred calls in capsules,
+//! this timer should be removed.
+//!
+//! Usage
+//! -----
+//!
+//! ```
+//! let virtual_alarm_rtt = static_init!(
+//!     capsules::virtual_alarm::VirtualMuxAlarm<'static, nrf5x::rtc::Rtc>,
+//!     capsules::virtual_alarm::VirtualMuxAlarm::new(mux_alarm)
+//! );
+//!
+//! let rtt_memory = static_init!(
+//!     capsules::segger_rtt::SeggerRttMemory,
+//!     capsules::segger_rtt::SeggerRttMemory::new(b"Terminal\0",
+//!         &mut capsules::segger_rtt::UP_BUFFER,
+//!         b"Terminal\0",
+//!         &mut capsules::segger_rtt::DOWN_BUFFER)
+//! );
+//!
+//! let rtt = static_init!(
+//!     capsules::segger_rtt::SeggerRtt<VirtualMuxAlarm<'static, nrf5x::rtc::Rtc>>,
+//!     capsules::segger_rtt::SeggerRtt::new(virtual_alarm_rtt, rtt_memory,
+//!         &mut capsules::segger_rtt::UP_BUFFER,
+//!         &mut capsules::segger_rtt::DOWN_BUFFER)
+//! );
+//! virtual_alarm_rtt.set_client(rtt);
+//! ```
+
+use kernel::common::cells::{OptionalCell, TakeCell};
+use kernel::hil;
+use kernel::hil::time::Frequency;
+
+/// Buffer for transmitting to the host.
+pub static mut UP_BUFFER: [u8; 1024] = [0; 1024];
+
+/// Buffer for receiving messages from the host.
+pub static mut DOWN_BUFFER: [u8; 32] = [0; 32];
+
+/// This structure is defined by the segger RTT protocol.
+#[repr(C)]
+pub struct SeggerRttMemory {
+    id: [u8; 16],
+    number_up_buffers: u32,
+    number_down_buffers: u32,
+    up_buffer: SeggerRttBuffer,
+    down_buffer: SeggerRttBuffer,
+}
+
+#[repr(C)]
+pub struct SeggerRttBuffer {
+    name: *const u8, // Pointer to the name of this channel. Must be a 4 byte thin pointer.
+    buffer: *const u8, // Pointer to the buffer for this channel.
+    length: u32,
+    write_position: u32,
+    read_position: u32,
+    flags: u32,
+}
+
+impl SeggerRttMemory {
+    pub fn new(
+        up_buffer_name: &'static [u8],
+        up_buffer: &'static [u8],
+        down_buffer_name: &'static [u8],
+        down_buffer: &'static [u8],
+    ) -> SeggerRttMemory {
+        SeggerRttMemory {
+            // Must be "SEGGER RTT".
+            id: [
+                0x53, 0x45, 0x47, 0x47, 0x45, 0x52, 0x20, 0x52, 0x54, 0x54, 0, 0, 0, 0, 0, 0,
+            ],
+            number_up_buffers: 1,
+            number_down_buffers: 1,
+            up_buffer: SeggerRttBuffer {
+                name: up_buffer_name.as_ptr(),
+                buffer: up_buffer.as_ptr(),
+                length: 1024,
+                write_position: 0,
+                read_position: 0,
+                flags: 0,
+            },
+            down_buffer: SeggerRttBuffer {
+                name: down_buffer_name.as_ptr(),
+                buffer: down_buffer.as_ptr(),
+                length: 32,
+                write_position: 0,
+                read_position: 0,
+                flags: 0,
+            },
+        }
+    }
+}
+
+pub struct SeggerRtt<'a, A: hil::time::Alarm + 'a> {
```

Comment:
docs, perhaps explain that `Receive` is not implemented yet?!

Reply:
What docs are you looking for? I don't see any docs on the main capsule struct in other capsules.

Reply:
It was because `_down_buffer` isn`t used yet but you have clarified that now in the `todo`